File: Liskovs/compliant/BankAccount.java
class BankAccount {
    protected double balance;

    public BankAccount(double initialBalance) {
        if (initialBalance < 0) throw new IllegalArgumentException();
        this.balance = initialBalance;
    }

    public void withdraw(double amount) {
        if (amount < 0 || amount>balance) throw new IllegalArgumentException();
        balance -= amount;
    }

    public double getBalance(){
        return balance;
    }
}

class SavingsAccount extends BankAccount {
    @Override
    public void withdraw(double amount) {
        super.withdraw(amount);
    }
}

File: Liskovs/compliant/BankAccount.ts
class BankAccount {
    protected balance: number;

    constructor(initialBalance: number) {
        if (initialBalance < 0) {
            throw new Error("Initial balance must be non-negative");
        }
        this.balance = initialBalance;
    }

    withdraw(amount: number): void {
        if (amount < 0 || amount > this.balance) {
            throw new Error("Invalid withdrawal amount");
        }
        this.balance -= amount;
    }

    getBalance(): number {
        return this.balance;
    }
}

class SavingsAccount extends BankAccount {
    withdraw(amount: number): void {
        // Preserves invariant
        super.withdraw(amount);
    }
}
